service-base/src/main/java/org/eclipse/hono/service/registration/AbstractRegistrationService.java

```diff
@@ -316,23 +360,32 @@ protected CacheDirective getRegistrationAssertionCacheDirective(final String dev
     /**
      * Gets the list of gateways that may act on behalf of the given device.
      * <p>
-     * This default implementation gets the list of gateways from the value of the
-     * {@link RegistrationConstants#FIELD_VIA} property in the device's registration information.
+     * To compile this list of gateways, this default implementation gets the list of gateways from the value of the
+     * {@link RegistrationConstants#FIELD_VIA} property in the device's registration information and resolves
+     * the members of the groups that are in the {@link RegistrationConstants#FIELD_VIA_GROUPS} property of the device.
      * <p>
      * Subclasses may override this method to provide a different means to determine the supported gateways.
      *
+     *
      * @param tenantId The tenant id.
      * @param deviceId The device id.
      * @param registrationInfo The device's registration information.
      * @return The list of gateways as a JSON array of Strings (never {@code null}).
      */
-    protected JsonArray getSupportedGatewaysForDevice(final String tenantId, final String deviceId,
+    protected Future<JsonArray> getSupportedGatewaysForDevice(final String tenantId, final String deviceId,
```

Comment:
How about adding a `span` parameter here (also in an additional variant of `getAssertionPayload`)? This could be used for the `resolveGroupMembers` invocation.